src/lib.rs

```diff
@@ -106,17 +129,33 @@ impl SimpleLogger {
         mut self,
         target_levels: HashMap<String, LevelFilter>,
     ) -> SimpleLogger {
-        self.module_levels = target_levels;
+        self.module_levels = target_levels.into_iter().collect();
+
+        /* Normally this is only called in `init` to avoid redundancy, but we can't initialize the logger in tests */
+        #[cfg(test)]
+        self.module_levels
+            .sort_by_key(|(name, _level)| name.len().wrapping_neg());
+
         self
     }
 
     /// 'Init' the actual logger, instantiate it and configure it,
     /// this method MUST be called in order for the logger to be effective.
-    pub fn init(self) -> Result<(), SetLoggerError> {
+    pub fn init(mut self) -> Result<(), SetLoggerError> {
         #[cfg(all(windows, feature = "colored"))]
         set_up_color_terminal();
 
-        let max_level = self.module_levels.values().copied().max();
+        /* Sort all module levels from most specific to least specific. The length of the module
+         * name is used instead of its actual depth to avoid module name parsing.
+         */
+        self.module_levels
+            .sort_by_key(|(name, _level)| name.len().wrapping_neg());
```

Comment:
Simple curiosity: why the use of `wrapping_neg ` ?

Reply:
I want to sort the list in reverse. I could have sorted it using a comparator function that does `|a, b| a.cmp(b).reversed()` or something but I preferred using the `sort_by_key` instead of writing a comparator.

`len()` gives us an `usize` which is obviously hard to negate. I first use `-(len as isize)` but that's rather ugly and then I found the `xxx_negate` methods on the type itself. I read the docs multiple times and I'm still not 100% sure about what `wrapping_negates` really does (especially because its written in a generic way that also applies to signed types). What I want is simply to map all values so that comparing them gives the reverse order.

Reply:
Got it for `wrapping_negates`, it is indeed weird to have this methods for an unsigned type, but looking at the documentation it will works.

But something came across my mind, I think ordering based on module name length isn't really robust and could lead to false positive, no?
Isn't a better idea to count the number of occurrences of the module separator "::" ?

Reply:
I'm 80% confident that this produces no false positives and I couldn't come up with any counter examples. Sketch of proof: If `a` is a sub-module of `b`, then its path is `b::a` per definition and thus longer.

Reply:
You are totally right, my bad!